### PR TITLE
fix: initialization updates

### PR DESCRIPTION
**Pull request by nperraud**
This PR attempt to fix:
- Initialisation problems in primal dual algorithms
- Function evaluation problem in primal dual algorithm

Additionally, maybe change the way that the initial value of the optimization algorithm is handled, i.e. that its variable will not be changed by the algorithm.

This PR is linked to #35 

closes: #35

### CODE REVIEW

**Comment by nperraud** [comment]
@mdeff Here is my argument why we should not modify the initial value provided by the algorithm.

- Why modify it, we save a bit of RAM.
- Why copy it, we might avoid a lot of stupid mistake on the user side.

I believe the second point is more important than the first one. So I suggest to add an option to modify it which is turned off by default.

**Comment by nperraud** [comment]
@mdeff Ready to merge. I might make other changes but I would prefer to make several PRs.

File: pyunlocbox/solvers.py [hunk 1]
```diff
@@ -780,14 +798,27 @@ def __init__(self, L=None, Lt=None, d0=None, *args, **kwargs):
...
+            # The copy is necessary in case `L = lambda x: x`.
+            self.dual_sol = self.L(np.asarray(x0).copy())
         else:
             self.dual_sol = self.d0
 
     def _post(self):
         self.d0 = None
         del self.dual_sol
 
+    def objective(self, x):
```
Comment (mdeff) [inline pyunlocbox/solvers.py:810]: Could we do as for the other methods, i.e., an `objective()` method with the documentation, that calls an `_objective()` implementation, which can be overloaded?

File: pyunlocbox/solvers.py [hunk 2]
```diff
@@ -54,7 +54,7 @@
 
 
 def solve(functions, x0, solver=None, atol=None, dtol=None, rtol=1e-3,
-          xtol=None, maxit=200, verbosity='LOW'):
+          xtol=None, maxit=200, verbosity='LOW', persistent=False):
```
Comment (mdeff) [inline pyunlocbox/solvers.py:57]: Is `persistent` a standard name for such kind of functionality?

**Review by mdeff — COMMENTED** [review]
Thanks for the fixes! And the tests. :)

**Comment by mdeff** [comment]
> I believe the second point is more important than the first one.

Strong agree.

> So I suggest to add an option to modify it which is turned off by default.

In general I prefer to avoid the proliferation of options. But I believe saving memory was important enough at some point for me to implement it in this way, so an option should be worth it.

File: pyunlocbox/solvers.py [hunk 3]
```diff
@@ -780,14 +798,27 @@ def __init__(self, L=None, Lt=None, d0=None, *args, **kwargs):
     def _pre(self, functions, x0):
         # Dual variable.
         if self.d0 is None:
-            self.dual_sol = self.L(x0)
+            # The copy is necessary in case `L = lambda x: x`.
```
[the remark this answers is not in the document]
Comment (nperraud) [reply pyunlocbox/solvers.py:801]: It does not change anything in the other cases.

[the remark this answers is not in the document]
Comment (nperraud) [reply pyunlocbox/solvers.py:801]: It was a bug, the function would not work if L is lambda x:x

File: pyunlocbox/solvers.py (hunk 2 above)
Comment (nperraud) [reply pyunlocbox/solvers.py]: I think not, I went with `inplace` instead.

File: pyunlocbox/solvers.py (hunk 1 above)
Comment (nperraud) [reply pyunlocbox/solvers.py]: Done

File: pyunlocbox/solvers.py [hunk 4]
```diff
@@ -862,7 +893,7 @@ def _algo(self):
         q1 = p1 - self.step * (self.smooth_funs[0].grad(p1) + self.Lt(p2))
         q2 = p2 + self.step * self.L(p1)
 
-        # Update solution (in both primal and dual spaces)
+        # Update solution (in both primal and dual spaces
```
[the remark this answers is not in the document]
Comment (nperraud) [reply pyunlocbox/solvers.py]: Fixed

**Comment by nperraud** [comment]
@mdeff Do I have you approval for merging?

File: pyunlocbox/solvers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdeff) [reply pyunlocbox/solvers.py:803]: > It does not change anything in the other cases.

Great.

File: pyunlocbox/solvers.py (hunk 1 above)
Comment (mdeff) [reply pyunlocbox/solvers.py]: Thanks!

File: pyunlocbox/solvers.py (hunk 2 above)
Comment (mdeff) [reply pyunlocbox/solvers.py]: I was going to suggest that name too. :)

**Comment by mdeff** [comment]
Thanks for the update. Though there's now an issue with your commit: it includes the additions of `proj_linalg` and `proj_sdsp` (maybe other things?) which we already merged.

**Comment by nperraud** [comment]
Sorry for that... I just rebased and now everything is good.

**Review by mdeff — APPROVED** [review]

**Comment by mdeff** [comment]
LGTM. Thanks!

(The second commit message could be more explicit but that's a detail.)

**Comment by nperraud** [comment]
🥳

**Pull request merged**